File: lib-segment/src/test/java/websiteschema/mpsegment/dict/PunctuationTest.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.dict;

import org.junit.Assert;
import org.junit.Test;

/**
 * @author ray
 */
public class PunctuationTest {

    private HashDictionary hashDictionary = null;

    public PunctuationTest() {
        DictionaryFactory.getInstance().loadDictionary();
        hashDictionary = DictionaryFactory.getInstance().getCoreDictionary();
    }

    @Test
    public void should_contains_basic_punctuations() {
        char punctuations[] = ",./<>?;':\"{}[]!@#$%^&*()_+-=\\|。，、；：？！…-·ˉˇ¨‘`~'“”々～‖∶＂＇｀｜〃〔〕〈〉《》「」『』．〖〗【】（）［］｛｝".toCharArray();
        for (char punctuation : punctuations) {
            System.out.print(punctuation + " ");
            IWord word = hashDictionary.getWord(String.valueOf(punctuation));
            Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_W);
        }
        System.out.println();
    }

    @Test
    public void should_contains_money_symbols() {
        char punctuations[] = "฿€￡₤￥".toCharArray();
        for (char punctuation : punctuations) {
            System.out.print(punctuation + " ");
            IWord word = hashDictionary.getWord(String.valueOf(punctuation));
            Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_Q);
        }
        System.out.println();
    }

    @Test
    public void should_contains_number_symbols() {
        char punctuations[] = ("ⅠⅡⅢⅣⅤⅥⅦⅧⅨⅩⅪⅫ" +
                "㈠㈡㈢㈣㈤㈥㈦㈧㈨㈩" +
                "⑴⑵⑶⑷⑸⑹⑺⑻⑼⑽" +
                "①②③④⑤⑥⑦⑧⑨⑩⑾⑿⒀⒁⒂⒃⒄⒅⒆⒇" +
                "①②③④⑤⑥⑦⑧⑨⑩№" +
                "⑴⑵⑶⑷⑸⑹⑺⑻⑼⑽⑾⑿⒀⒁⒂⒃⒄⒅⒆⒇" +
                "⒈⒉⒊⒋⒌⒍⒎⒏⒐⒑⒒⒓⒔⒕⒖⒗⒘⒙⒚⒛" +
                "ⅠⅡⅢⅣⅤⅥⅦⅧⅨⅩⅪⅫⅰⅱⅲⅳⅴⅵⅶⅷⅸⅹ").toCharArray();
        for (char punctuation : punctuations) {
            System.out.print(punctuation + " ");
            IWord word = hashDictionary.getWord(String.valueOf(punctuation));
            Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_M);
        }
        System.out.println();
    }

    @Test
    public void should_contains_math_symbols() {
        char punctuations[] = ("≈≡≠＝≤≥＜＞≮≯∷±＋－×÷／∫∮∝∞∧∨∑∏∪∩∈∵∴⊥∥∠⌒⊙≌∽√").toCharArray();
        for (char punctuation : punctuations) {
            System.out.print(punctuation + " ");
            IWord word = hashDictionary.getWord(String.valueOf(punctuation));
            Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_W);
        }
        System.out.println();
    }

    @Test
    public void should_contains_units_of_measurement() {
        char punctuations[] = ("㎎㎏㎜㎝㎞㎡㏄㏎㏑°′″＄￡￥‰％℃¤￠").toCharArray();
        for (char punctuation : punctuations) {
            System.out.print(punctuation + " ");
            IWord word = hashDictionary.getWord(String.valueOf(punctuation));
            Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_Q);
        }
        System.out.println();
    }

    @Test
    public void should_contains_breaking_characters() {
        String punctuations[] = new String[]{"\r\n", "\r", "\n", " ", "　"};
        for (String punctuation : punctuations) {
            System.out.print(java.net.URLEncoder.encode(punctuation) + " ");
            IWord word = hashDictionary.getWord(punctuation);
            Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_W);
        }
        System.out.println();
    }

    @Test
    public void should_contains_other_characters() {
        char punctuations[] = ("┌┍┎┏┐┑┒┓─┄┈├┝┞┟┠┡┢┣│┆┊┬┭┮┯┰┱┲┳┼┽┾┿╀╁╂╃§☆★●◎◇◆□■△▲※→←↑↓〓＃＿＆＠＼＾αβγδεζηθικλμνξοπρστυφχψωΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩабвгдеёжзийклмнопрстуфхцчшщъыьэюяАБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯㄅㄉㄓㄚㄞㄢㄦㄆㄊㄍㄐㄔㄗㄧㄛㄟㄣㄇㄋㄎㄑㄕㄘㄨㄜㄠㄤㄈㄏㄒㄖㄙㄩㄝㄡㄥāáǎàōóǒòêēéěèīíǐìūúǔùǖǘǚǜぁぃぅぇぉかきくけこんさしすせそたちつってとゐなにぬねのはひふへほゑまみむめもゃゅょゎをあいうえおがぎぐげござじずぜぞだぢづでどぱぴぷぺぽぼびぶべぼらりるれろやゆよわァィゥヴェォカヵキクケヶコサシスセソタチツッテトヰンナニヌネノハヒフヘホヱマミムメモャュョヮヲアイウエオガギグゲゴザジズゼゾダヂヅデドパピプペポバビブベボラリルレロヤユヨワ]レロヤユヨワ").toCharArray();
        for (char punctuation : punctuations) {
            System.out.print(punctuation + " ");
            IWord word = hashDictionary.getWord(String.valueOf(punctuation));
            Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_W);
        }
        System.out.println();
    }

    private int getFirstPosOfWord(IWord word) {
        return word.getPOSArray().getWordPOSTable()[0][0];
    }
}

File: lib-segment/src/test/scala/me/yingrui/segment/dict/PunctuationTest.scala
package me.yingrui.segment.dict

import org.junit.Assert
import org.junit.Assert._
import org.junit.Test

class PunctuationTest {

  DictionaryFactory().loadDictionary()
  private val hashDictionary = DictionaryFactory().getCoreDictionary

  @Test
  def should_contains_basic_punctuations() {
    val punctuations = ",./<>?;':\"{}[]!@#$%^&*()_+-=\\|。，、；：？！…-·ˉˇ¨‘`~'“”々～‖∶＂＇｀｜〃〔〕〈〉《》「」『』．〖〗【】（）［］｛｝".toCharArray()
    for (punctuation <- punctuations) {
      val word = hashDictionary.getWord(punctuation.toString)
      print(punctuation + "," + getFirstPosOfWord(word) + "; ")
      assertEquals(POSUtil.POS_W, getFirstPosOfWord(word))
    }
    println()
  }

  @Test
  def should_contains_money_symbols() {
    val punctuations = "฿€￡₤￥".toCharArray()
    for (punctuation <- punctuations) {
      print(punctuation + " ")
      val word = hashDictionary.getWord(punctuation.toString)
      Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_Q)
    }
    println()
  }

  @Test
  def should_contains_number_symbols() {
    val punctuations = ("ⅠⅡⅢⅣⅤⅥⅦⅧⅨⅩⅪⅫ" +
      "㈠㈡㈢㈣㈤㈥㈦㈧㈨㈩" +
      "⑴⑵⑶⑷⑸⑹⑺⑻⑼⑽" +
      "①②③④⑤⑥⑦⑧⑨⑩⑾⑿⒀⒁⒂⒃⒄⒅⒆⒇" +
      "①②③④⑤⑥⑦⑧⑨⑩№" +
      "⑴⑵⑶⑷⑸⑹⑺⑻⑼⑽⑾⑿⒀⒁⒂⒃⒄⒅⒆⒇" +
      "⒈⒉⒊⒋⒌⒍⒎⒏⒐⒑⒒⒓⒔⒕⒖⒗⒘⒙⒚⒛" +
      "ⅠⅡⅢⅣⅤⅥⅦⅧⅨⅩⅪⅫⅰⅱⅲⅳⅴⅵⅶⅷⅸⅹ").toCharArray()
    for (punctuation <- punctuations) {
      print(punctuation + " ")
      val word = hashDictionary.getWord(punctuation.toString)
      Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_M)
    }
    println()
  }

  @Test
  def should_contains_math_symbols() {
    val punctuations = ("≈≡≠＝≤≥＜＞≮≯∷±＋－×÷／∫∮∝∞∧∨∑∏∪∩∈∵∴⊥∥∠⌒⊙≌∽√").toCharArray()
    for (punctuation <- punctuations) {
      print(punctuation + " ")
      val word = hashDictionary.getWord(punctuation.toString)
      Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_W)
    }
    println()
  }

  @Test
  def should_contains_units_of_measurement() {
    val punctuations = ("㎎㎏㎜㎝㎞㎡㏄㏎㏑°′″＄￡￥‰％℃¤￠").toCharArray()
    for (punctuation <- punctuations) {
      print(punctuation + " ")
      val word = hashDictionary.getWord(punctuation.toString)
      Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_Q)
    }
    println()
  }

  @Test
  def should_contains_breaking_characters() {
    val punctuations = List(
      "\r\n", "\r", "\n", " ", "　"
    )
    for (punctuation <- punctuations) {
      print(java.net.URLEncoder.encode(punctuation) + " ")
      val word = hashDictionary.getWord(punctuation)
      Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_W)
    }
    println()
  }

  @Test
  def should_contains_other_characters() {
    val punctuations = ("┌┍┎┏┐┑┒┓─┄┈├┝┞┟┠┡┢┣│┆┊┬┭┮┯┰┱┲┳┼┽┾┿╀╁╂╃§☆★●◎◇◆□■△▲※→←↑↓〓＃＿＆＠＼＾αβγδεζηθικλμνξοπρστυφχψωΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩабвгдеёжзийклмнопрстуфхцчшщъыьэюяАБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯㄅㄉㄓㄚㄞㄢㄦㄆㄊㄍㄐㄔㄗㄧㄛㄟㄣㄇㄋㄎㄑㄕㄘㄨㄜㄠㄤㄈㄏㄒㄖㄙㄩㄝㄡㄥāáǎàōóǒòêēéěèīíǐìūúǔùǖǘǚǜぁぃぅぇぉかきくけこんさしすせそたちつってとゐなにぬねのはひふへほゑまみむめもゃゅょゎをあいうえおがぎぐげござじずぜぞだぢづでどぱぴぷぺぽぼびぶべぼらりるれろやゆよわァィゥヴェォカヵキクケヶコサシスセソタチツッテトヰンナニヌネノハヒフヘホヱマミムメモャュョヮヲアイウエオガギグゲゴザジズゼゾダヂヅデドパピプペポバビブベボラリルレロヤユヨワ]レロヤユヨワ").toCharArray()
    for (punctuation <- punctuations) {
      print(punctuation + " ")
      val word = hashDictionary.getWord(punctuation.toString)
      Assert.assertEquals(getFirstPosOfWord(word), POSUtil.POS_W)
    }
    println()
  }

  private def getFirstPosOfWord(word: IWord) = word.getPOSArray().getWordPOSTable()(0)(0)
}
